exe/tests/helpers/mod.rs

```diff
@@ -1,19 +1,87 @@
 use std::ffi::OsStr;
-use std::io;
+use std::fs;
 use std::path::Path;
+use std::path::PathBuf;
 use std::process::Command;
 use std::str;
 
 use anyhow::bail;
 use assert_cmd::cargo::CommandCargoExt;
+use std::env;
 
-pub fn get_image<P: AsRef<Path>>(to_dir: P) -> io::Result<()> {
+pub fn get_image<P: AsRef<Path>>(to_dir: P) -> anyhow::Result<PathBuf> {
     let image = "ubuntu";
     let tag = "latest";
     if to_dir.as_ref().exists() {
-        return Ok(());
+        let rootfs = to_dir.as_ref().join("rootfs");
+        if rootfs.exists() {
+            return Ok(rootfs);
+        } else {
+            bail!(
+                "{:?} exists but does not have a rootfs directory in it",
+                to_dir.as_ref().display()
+            );
+        }
     }
-    docker_extract::extract_image(image, tag, to_dir.as_ref())
+
+    let mut xdg_data_home_default = env::var("HOME").unwrap();
+    xdg_data_home_default.push_str("/.local/share");
+
+    let mut xdg_data_home = env::var("XDG_DATA_HOME").unwrap_or(xdg_data_home_default.clone());
+    if xdg_data_home.is_empty() {
+        xdg_data_home = xdg_data_home_default;
+    }
+
+    let puzzlefs_data_path = format!("{xdg_data_home}/puzzlefs");
+    fs::create_dir_all(&puzzlefs_data_path)?;
+
+    // skopeo copy docker://ubuntu:latest oci:$HOME/.local/share/puzzlefs/ubuntu:latest
+    let output = Command::new("skopeo")
+        .args([
+            "copy",
+            &format!("docker://{image}:{tag}"),
+            &format!("oci:{puzzlefs_data_path}/{image}:{tag}"),
+        ])
+        .output()?;
+    if !output.status.success() {
+        bail!(
+            "skopeo exited with error:\n{}",
+            str::from_utf8(&output.stderr)?,
+        );
+    }
+
+    if !output.stdout.is_empty() {
+        println!(
+            "skopeo output\n{}",
+            str::from_utf8(&output.stdout).expect("Script output should not contain non-UTF8")
+        );
+    }
+
+    // umoci unpack --rootless --image ubuntu:latest /tmp/.tmpxyz/ubuntu
```

Comment:
This comment however is wrong.  As is, you're not unpacking "ubuntu:latest", but simply an image called "latest".

Reply:
This is what gets executed: `umoci unpack --rootless --image ~/.local/share/puzzlefs/ubuntu:latest /tmp/.tmpxyz/ubuntu`, and in the comment I didn't specify the full path, only `ubuntu`.

`umoci unpack --help` says:
```
USAGE:
   umoci unpack [command options] --image <image-path>[:<tag>] <bundle>

Where "<image-path>" is the path to the OCI image, "<tag>" is the name of the
tagged image to unpack (if not specified, defaults to "latest") and "<bundle>"
is the destination to unpack the image to.
```
So is my comment wrong because:
* I'm unpacking `image` which is `ubuntu` for now but might change in the future?
* I didn't specify the full path to the `ubuntu` directory?
* other reason?

Reply:
Yeah I don't know why i thought it was wrong.